compiler.ts

```diff
@@ -451,38 +452,112 @@ function codeGenDestructure(
+        assignStmts = destruct.targets.flatMap((target) => {
+          const assignable = target.target;
+          if (target.starred) {
+            throw new Error("Do not currently support starred assignment targets");
+          } else {
+            const [offset, _] = classFields.next().value;
+            // The WASM code value that we extracted from the object at this current offset
+            const addressOffset = offset * 4;
+            const fieldValue = [value, `(i32.load offset=${addressOffset})`];
+            return codeGenAssignable(assignable, fieldValue, env);
+          }
         });
         break;
       }
       case "tuple": {
         let offset = 0;
-        for (let target of destruct.targets) {
+
+        assignStmts = destruct.targets.flatMap((target) => {
+          const assignable = target.target;
           if (target.starred) {
             throw new Error("Do not currently support starred assignment targets");
-          } else if (!target.ignore) {
-            let fieldValue = [value, `(i32.load offset=${offset})`];
-            assignStmts.push(...codeGenAssignable(target.target, fieldValue, env));
+          } else {
+            const fieldValue = [value, `(i32.load offset=${offset})`];
             offset += 4;
+            return codeGenAssignable(assignable, fieldValue, env);
           }
+        });
+        break;
+      }
+      case "list": {
+        // Determines if we have a single target with starred == true
+        // used to run a different runtime length check
+        const isStarred = destruct.targets.some((target) => target.starred);
+
+        let targetListLength: number;
+        let compareLengthOperator: string;
+        if (isStarred) {
+          // Targets list length - 1 is the lower bound for number of required elements in our array
+          // The array can be larger/equal to this (because of star operator) but not smaller
+          targetListLength = destruct.targets.length - 1;
+          compareLengthOperator = `(i32.ge_u)`;
+        } else {
+          // No star operator simply means the target lists and the runtime list must be same length
+          targetListLength = destruct.targets.length;
+          compareLengthOperator = `(i32.eq)`;
         }
+
+        // TODO: Add nice runtime error instead of "RuntimeError: unreachable"
+        const lengthCheck = [
+          value,
+          `(i32.load offset=4)`, // list length, stored at byte offset 4
+          `(i32.const ${targetListLength})`, // number of destruct targets
+          compareLengthOperator,
+          `(if (then (nop)) (else (unreachable)))`,
+          //`(if (then (nop)) (else (call $SomeExitingErrorFunction)))`
+        ];
+
+        // Skip past the three header values of lists and to the actual values
+        let offset = 12;
+
+        const targetStmts = destruct.targets.flatMap((target) => {
+          const assignable = target.target;
+          if (target.starred) {
```

Comment:
I think I'd feel much safer if the "copy" logic here were broken out into a separate method since it's more "list datastructure" focused code and less about the actual destructuring. It's really close to being able to use the existing `codeGenListCopy`, but not quite. might be worth talking with lists to see if that could be modified slightly for this use case (copying a subset of a list) so we don't have to duplicate logic as much.

Reply:
Hmm okay good idea, I was looking for some way to not have to write so much of this list code and that would be a very nice solution, didn't realize how close ```codeGenListCopy``` is to our use case. I'll ask in the slack about it